keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java

```diff
@@ -158,37 +163,24 @@ public PartialDerivatives subtract(PartialDerivatives subtraction, long[] ofShap
         return new PartialDerivatives(subtracted);
     }
 
-    private static Map<VertexId, DoubleTensor> cloneWithCorrectShape(Map<VertexId, DoubleTensor> infinitesimals, long[] ofShape) {
+    private static Map<VertexId, DoubleTensor> cloneWithCorrectShape(Map<VertexId, DoubleTensor> infinitesimals,
+                                                                     boolean ofIsScalar,
+                                                                     long[] resultShape) {
 
         Map<VertexId, DoubleTensor> clone = new HashMap<>();
         for (Map.Entry<VertexId, DoubleTensor> entry : infinitesimals.entrySet()) {
             VertexId k = entry.getKey();
             DoubleTensor v = entry.getValue();
 
-            if (ofShape == null || ofShapeMatches(ofShape, v.getShape())) {
-                clone.put(k, v);
+            if (ofIsScalar) {
```

Comment:
Nope. Scalar is rank 0. Length one is any [1,1], [1,1,1] or [] as well. All scalars are length one but not all length ones are scalar. It's a special case because length one is always broadcastable.